EIPS/eip-7928.md

```diff
@@ -0,0 +1,199 @@
+Each entry MUST include the transaction index and the signed balance delta per address for each transaction.
+11 bytes are sufficient to represent the total ETH supply.
+
+The `NonceDiffs` structure MUST record the post-block nonce values for all `CREATE2` deployer accounts and the deploayed contracts in the block.
+
+### State Transition Function
+
+Modify the state transition function to validate the block-level access lists:
+
+```python 
+def state_transition(block: Block) -> None:
+    computed_access_list = {}
+    computed_balance_diffs = {}
+    computed_nonce_diffs = {}
+
+    for idx, tx in enumerate(block.transactions):
+        # Execute transaction and track state accesses
+        accessed_items, balance_changes = execute_transaction(tx)
+        
+        # Process storage accesses
+        for (address, slot), is_write, value in accessed_items:
+            if (address, slot) not in computed_access_list:
+                computed_access_list[(address, slot)] = []
+            
+            access_entry = {"tx_index": idx}
+            if is_write:
+                access_entry["value_after"] = value
+            
+            computed_access_list[(address, slot)].append(access_entry)
+        
+        # Process balance changes
+        for address, delta in balance_changes:
+            if address not in computed_balance_diffs:
+                computed_balance_diffs[address] = []
+            computed_balance_diffs[address].append({"tx_index": idx, "delta": delta})
+
+    # Validate access list and balance diffs
+    if not validate_access_list(block.block_access_list, computed_access_list):
+        raise InvalidBlock("Mismatch in block-level access list.")
+        
+    if not validate_balance_diffs(block.balance_diffs, computed_balance_diffs):
+        raise InvalidBlock("Mismatch in balance diffs.")
+        
+    if not validate_nonce_diffs(block.nonce_diffs, computed_nonce_diffs):
+        raise InvalidBlock("Mismatch in nonce diffs.")
+```
+
+The BAL MUST be complete and accurate. It MUST NOT contain too few entries (missing accesses) or too many entries (spurious accesses). Any missing or extra entries in the access list, balance diffs, or nonce diffs SHALL result in block invalidation.
+
+Client implementations MUST compare the accessed addresses and storage keys gathered during execution (as defined in [EIP-2929](./eip-2929.md)) with those included in the BAL to determine validity.
+
+Client implementations MAY invalidate the block right away in case any transaction steps outside the declared state.
+
+## Rationale
+
+### BAL Design Choice
+
+We chose the BAL variant for several key reasons:
+
+1. **Balance between size and parallelization benefits**: BALs enable both parallel disk reads and parallel EVM execution while keeping the size manageable.
+
+2. **Storage value inclusion**: Including post-execution values for write operations facilitates state reconstruction during syncing, enabling faster chain catch-up.
```

Comment:
Wait, WOW this is actually a super good point! So this is super related to the diff layer of the snap sync! Wow I had not realized this, will keep this in mind, maybe most of the logic we want here is already available at the devp2p layer (snap sync).

Note: the snap sync diff layer only diffs at the block level, not at tx level. So with the snap sync diff layer we do not get the benefit of the tx parallelization.

Reply:
Yeah right, the only thing that snap sync gives you that such state diffs don't is a proof that the updated state corresponds to the root in the block. However, we don't get actual execution proofs.
The state diffs can be applied during sync and we need at least post-block values for catching up.